File: 2023/str-230417_0512/4889.cpp
#include <cstdio>

int cost(char s[])
{
    int inv = 0;
    int bias = 0;
    for (int i = 0; s[i]; i++) {
        if (s[i] == '}')
            bias--;
        else if (s[i] == '{')
            bias++;
        
        if (bias == -1) {
            bias = 1;
            inv++;
        }
    }
    return inv + bias / 2;
}

char s[2001];

int main()
{
    for (int i = 1; ; i++) {
        scanf("%s", s);
        if (s[0] == '-')
            break;
        printf("%d. %d\n", i, cost(s));
    }
}

File: boj/2023/str-230417_0512/4889.py
import sys

def cost(s):
    inv = 0
    bias = 0

    '''
    while len(s) > 0:
        c, *s = s   # 이게 더 느림 (264ms vs 48ms)
        if c == '}':
            bias -= 1
        elif c == '{':
            bias += 1
        
        if bias == -1:
            bias = 1
            inv += 1
    '''

    for i in range(len(s)):
        if s[i] == '}':
            bias -= 1
        elif s[i] == '{':
            bias += 1
        
        if bias == -1:
            bias = 1
            inv += 1
    
    return inv + int(bias / 2)

for i, s in enumerate(sys.stdin.readlines()):
    s = s.rstrip()

    if s[0] == '-':
        break

    print(f"{i + 1}. {cost(s)}")